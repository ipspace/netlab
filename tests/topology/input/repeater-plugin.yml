---
plugin: [ repeater ]

module: [lag,vlan]

vlans:
 red:
  links: [r-h]  # VLAN links get updated

groups:
 hosts:
  members: [h]  # Gets updated with cloned nodes

nodes:
 r:
  device: frr

 h:
  device: frr
  repeat: 3     # Create 3 copies -> 4 hosts total
  id: 10        # Starting at ID 10

links:
- r-h           # Plugin supports different styles of specifying links, all normalized
- r:
    ifindex: 20 # Custom port on router -> gets incremented
  h:
    mtu: 1600   # Interface atts get copied too

- lag.members: [ r-h, r-h ] # TODO lags need special work, not supported yet -> plugin issues a warning and doesn't modify them