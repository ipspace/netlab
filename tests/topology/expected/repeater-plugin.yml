groups:
  hosts:
    members:
    - h
    - h-2
    - h-3
    - h-4
input:
- topology/input/repeater-plugin.yml
- package:topology-defaults.yml
lag:
  lacp: fast
  lacp_mode: active
  mode: 802.3ad
links:
- _linkname: links[1]
  interfaces:
  - ifindex: 1
    ifname: eth1
    ipv4: 10.1.0.2/30
    node: r
  - ifindex: 1
    ifname: eth1
    ipv4: 10.1.0.1/30
    node: h
  linkindex: 1
  node_count: 2
  prefix:
    ipv4: 10.1.0.0/30
  type: p2p
- _linkname: links[2]
  interfaces:
  - ifindex: 20
    ifname: eth20
    ipv4: 10.1.0.6/30
    node: r
  - ifindex: 2
    ifname: eth2
    ipv4: 10.1.0.5/30
    mtu: 1600
    node: h
  linkindex: 2
  node_count: 2
  prefix:
    ipv4: 10.1.0.4/30
  type: p2p
- _linkname: links[3]
  bridge: input_3
  interfaces:
  - ifindex: 30000
    ifname: bond1
    ipv4: 10.1.0.9/30
    node: r
  - ifindex: 30000
    ifname: bond1
    ipv4: 10.1.0.10/30
    node: h
  lag:
    ifindex: 1
  linkindex: 3
  node_count: 2
  prefix:
    ipv4: 10.1.0.8/30
  type: lag
- _linkname: links[4]
  interfaces:
  - ifindex: 1
    ifname: eth1
    ipv4: 10.1.0.13/30
    node: h-2
  - ifindex: 2
    ifname: eth2
    ipv4: 10.1.0.14/30
    node: r
  linkindex: 4
  node_count: 2
  prefix:
    ipv4: 10.1.0.12/30
  type: p2p
- _linkname: links[5]
  interfaces:
  - ifindex: 1
    ifname: eth1
    ipv4: 10.1.0.17/30
    node: h-3
  - ifindex: 3
    ifname: eth3
    ipv4: 10.1.0.18/30
    node: r
  linkindex: 5
  node_count: 2
  prefix:
    ipv4: 10.1.0.16/30
  type: p2p
- _linkname: links[6]
  interfaces:
  - ifindex: 1
    ifname: eth1
    ipv4: 10.1.0.21/30
    node: h-4
  - ifindex: 4
    ifname: eth4
    ipv4: 10.1.0.22/30
    node: r
  linkindex: 6
  node_count: 2
  prefix:
    ipv4: 10.1.0.20/30
  type: p2p
- _linkname: links[7]
  interfaces:
  - ifindex: 2
    ifname: eth2
    ipv4: 10.1.0.25/30
    mtu: 1600
    node: h-2
  - ifindex: 21
    ifname: eth21
    ipv4: 10.1.0.26/30
    node: r
  linkindex: 7
  node_count: 2
  prefix:
    ipv4: 10.1.0.24/30
  type: p2p
- _linkname: links[8]
  interfaces:
  - ifindex: 2
    ifname: eth2
    ipv4: 10.1.0.29/30
    mtu: 1600
    node: h-3
  - ifindex: 22
    ifname: eth22
    ipv4: 10.1.0.30/30
    node: r
  linkindex: 8
  node_count: 2
  prefix:
    ipv4: 10.1.0.28/30
  type: p2p
- _linkname: links[9]
  interfaces:
  - ifindex: 2
    ifname: eth2
    ipv4: 10.1.0.33/30
    mtu: 1600
    node: h-4
  - ifindex: 23
    ifname: eth23
    ipv4: 10.1.0.34/30
    node: r
  linkindex: 9
  node_count: 2
  prefix:
    ipv4: 10.1.0.32/30
  type: p2p
- _linkname: links[3].lag[1]
  interfaces:
  - ifindex: 5
    ifname: eth5
    node: r
  - ifindex: 3
    ifname: eth3
    node: h
  lag:
    _parentindex: 3
  linkindex: 10
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[3].lag[2]
  interfaces:
  - ifindex: 6
    ifname: eth6
    node: r
  - ifindex: 4
    ifname: eth4
    node: h
  lag:
    _parentindex: 3
  linkindex: 11
  node_count: 2
  prefix: false
  type: p2p
- _linkname: vlans.red.links[1]
  bridge: input_12
  interfaces:
  - _vlan_mode: irb
    ifindex: 7
    ifname: eth7
    ipv4: 172.16.0.1/24
    node: r
    vlan:
      access: red
  - _vlan_mode: irb
    ifindex: 5
    ifname: eth5
    ipv4: 172.16.0.10/24
    node: h
    vlan:
      access: red
  linkindex: 12
  node_count: 2
  prefix:
    allocation: id_based
    ipv4: 172.16.0.0/24
  type: lan
  vlan:
    access: red
- _linkname: vlans.red.links[2]
  bridge: input_13
  interfaces:
  - _vlan_mode: irb
    ifindex: 3
    ifname: eth3
    ipv4: 172.16.0.11/24
    node: h-2
    vlan:
      access: red
  - _vlan_mode: irb
    ifindex: 8
    ifname: eth8
    ipv4: 172.16.0.1/24
    node: r
    vlan:
      access: red
  linkindex: 13
  node_count: 2
  prefix:
    allocation: id_based
    ipv4: 172.16.0.0/24
  type: lan
  vlan:
    access: red
- _linkname: vlans.red.links[3]
  bridge: input_14
  interfaces:
  - _vlan_mode: irb
    ifindex: 3
    ifname: eth3
    ipv4: 172.16.0.12/24
    node: h-3
    vlan:
      access: red
  - _vlan_mode: irb
    ifindex: 9
    ifname: eth9
    ipv4: 172.16.0.1/24
    node: r
    vlan:
      access: red
  linkindex: 14
  node_count: 2
  prefix:
    allocation: id_based
    ipv4: 172.16.0.0/24
  type: lan
  vlan:
    access: red
- _linkname: vlans.red.links[4]
  bridge: input_15
  interfaces:
  - _vlan_mode: irb
    ifindex: 3
    ifname: eth3
    ipv4: 172.16.0.13/24
    node: h-4
    vlan:
      access: red
  - _vlan_mode: irb
    ifindex: 10
    ifname: eth10
    ipv4: 172.16.0.1/24
    node: r
    vlan:
      access: red
  linkindex: 15
  node_count: 2
  prefix:
    allocation: id_based
    ipv4: 172.16.0.0/24
  type: lan
  vlan:
    access: red
module:
- lag
- vlan
name: input
nodes:
  h:
    af:
      ipv4: true
    box: generic/ubuntu2004
    device: frr
    id: 10
    interfaces:
    - ifindex: 1
      ifname: eth1
      ipv4: 10.1.0.1/30
      linkindex: 1
      name: h -> r
      neighbors:
      - ifname: eth1
        ipv4: 10.1.0.2/30
        node: r
      type: p2p
    - ifindex: 2
      ifname: eth2
      ipv4: 10.1.0.5/30
      linkindex: 2
      mtu: 1600
      name: h -> r
      neighbors:
      - ifname: eth20
        ipv4: 10.1.0.6/30
        node: r
      type: p2p
    - ifindex: 30000
      ifname: bond1
      ipv4: 10.1.0.10/30
      lag:
        ifindex: 1
        lacp: fast
        lacp_mode: active
        mode: 802.3ad
      linkindex: 3
      name: h -> r
      neighbors:
      - ifname: bond1
        ipv4: 10.1.0.9/30
        node: r
      type: lag
      virtual_interface: true
    - ifindex: 3
      ifname: eth3
      lag:
        _parentindex: 3
      linkindex: 10
      name: h -> r
      neighbors:
      - ifname: eth5
        node: r
      type: p2p
    - ifindex: 4
      ifname: eth4
      lag:
        _parentindex: 3
      linkindex: 11
      name: h -> r
      neighbors:
      - ifname: eth6
        node: r
      type: p2p
    - bridge: input_12
      ifindex: 5
      ifname: eth5
      linkindex: 12
      name: '[Access VLAN red] h -> r'
      neighbors:
      - ifname: eth7
        ipv4: 172.16.0.1/24
        node: r
      type: lan
      vlan:
        access: red
        access_id: 1000
    - bridge_group: 1
      ifindex: 6
      ifname: vlan1000
      ipv4: 172.16.0.10/24
      name: VLAN red (1000) -> [r,h-2,h-3,h-4]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.1/24
        node: r
      - ifname: vlan1000
        ipv4: 172.16.0.11/24
        node: h-2
      - ifname: vlan1000
        ipv4: 172.16.0.12/24
        node: h-3
      - ifname: vlan1000
        ipv4: 172.16.0.13/24
        node: h-4
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
      mode: 802.3ad
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.10/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.110
      mac: 08:4f:a9:00:00:0a
    module:
    - lag
    - vlan
    name: h
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  h-2:
    af:
      ipv4: true
    box: generic/ubuntu2004
    device: frr
    id: 11
    interfaces:
    - ifindex: 1
      ifname: eth1
      ipv4: 10.1.0.13/30
      linkindex: 4
      name: h-2 -> r
      neighbors:
      - ifname: eth2
        ipv4: 10.1.0.14/30
        node: r
      type: p2p
    - ifindex: 2
      ifname: eth2
      ipv4: 10.1.0.25/30
      linkindex: 7
      mtu: 1600
      name: h-2 -> r
      neighbors:
      - ifname: eth21
        ipv4: 10.1.0.26/30
        node: r
      type: p2p
    - bridge: input_13
      ifindex: 3
      ifname: eth3
      linkindex: 13
      name: '[Access VLAN red] h-2 -> r'
      neighbors:
      - ifname: eth8
        ipv4: 172.16.0.1/24
        node: r
      type: lan
      vlan:
        access: red
        access_id: 1000
    - bridge_group: 1
      ifindex: 4
      ifname: vlan1000
      ipv4: 172.16.0.11/24
      name: VLAN red (1000) -> [h,r,h-3,h-4]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.10/24
        node: h
      - ifname: vlan1000
        ipv4: 172.16.0.1/24
        node: r
      - ifname: vlan1000
        ipv4: 172.16.0.12/24
        node: h-3
      - ifname: vlan1000
        ipv4: 172.16.0.13/24
        node: h-4
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
      mode: 802.3ad
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.11/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.111
      mac: 08:4f:a9:00:00:0b
    module:
    - lag
    - vlan
    name: h-2
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  h-3:
    af:
      ipv4: true
    box: generic/ubuntu2004
    device: frr
    id: 12
    interfaces:
    - ifindex: 1
      ifname: eth1
      ipv4: 10.1.0.17/30
      linkindex: 5
      name: h-3 -> r
      neighbors:
      - ifname: eth3
        ipv4: 10.1.0.18/30
        node: r
      type: p2p
    - ifindex: 2
      ifname: eth2
      ipv4: 10.1.0.29/30
      linkindex: 8
      mtu: 1600
      name: h-3 -> r
      neighbors:
      - ifname: eth22
        ipv4: 10.1.0.30/30
        node: r
      type: p2p
    - bridge: input_14
      ifindex: 3
      ifname: eth3
      linkindex: 14
      name: '[Access VLAN red] h-3 -> r'
      neighbors:
      - ifname: eth9
        ipv4: 172.16.0.1/24
        node: r
      type: lan
      vlan:
        access: red
        access_id: 1000
    - bridge_group: 1
      ifindex: 4
      ifname: vlan1000
      ipv4: 172.16.0.12/24
      name: VLAN red (1000) -> [h,r,h-2,h-4]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.10/24
        node: h
      - ifname: vlan1000
        ipv4: 172.16.0.1/24
        node: r
      - ifname: vlan1000
        ipv4: 172.16.0.11/24
        node: h-2
      - ifname: vlan1000
        ipv4: 172.16.0.13/24
        node: h-4
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
      mode: 802.3ad
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.12/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.112
      mac: 08:4f:a9:00:00:0c
    module:
    - lag
    - vlan
    name: h-3
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  h-4:
    af:
      ipv4: true
    box: generic/ubuntu2004
    device: frr
    id: 13
    interfaces:
    - ifindex: 1
      ifname: eth1
      ipv4: 10.1.0.21/30
      linkindex: 6
      name: h-4 -> r
      neighbors:
      - ifname: eth4
        ipv4: 10.1.0.22/30
        node: r
      type: p2p
    - ifindex: 2
      ifname: eth2
      ipv4: 10.1.0.33/30
      linkindex: 9
      mtu: 1600
      name: h-4 -> r
      neighbors:
      - ifname: eth23
        ipv4: 10.1.0.34/30
        node: r
      type: p2p
    - bridge: input_15
      ifindex: 3
      ifname: eth3
      linkindex: 15
      name: '[Access VLAN red] h-4 -> r'
      neighbors:
      - ifname: eth10
        ipv4: 172.16.0.1/24
        node: r
      type: lan
      vlan:
        access: red
        access_id: 1000
    - bridge_group: 1
      ifindex: 4
      ifname: vlan1000
      ipv4: 172.16.0.13/24
      name: VLAN red (1000) -> [h,r,h-2,h-3]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.10/24
        node: h
      - ifname: vlan1000
        ipv4: 172.16.0.1/24
        node: r
      - ifname: vlan1000
        ipv4: 172.16.0.11/24
        node: h-2
      - ifname: vlan1000
        ipv4: 172.16.0.12/24
        node: h-3
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
      mode: 802.3ad
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.13/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.113
      mac: 08:4f:a9:00:00:0d
    module:
    - lag
    - vlan
    name: h-4
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  r:
    _set_ifindex: true
    af:
      ipv4: true
    box: generic/ubuntu2004
    device: frr
    id: 1
    interfaces:
    - ifindex: 1
      ifname: eth1
      ipv4: 10.1.0.2/30
      linkindex: 1
      name: r -> h
      neighbors:
      - ifname: eth1
        ipv4: 10.1.0.1/30
        node: h
      type: p2p
    - ifindex: 2
      ifname: eth2
      ipv4: 10.1.0.14/30
      linkindex: 4
      name: r -> h-2
      neighbors:
      - ifname: eth1
        ipv4: 10.1.0.13/30
        node: h-2
      type: p2p
    - ifindex: 3
      ifname: eth3
      ipv4: 10.1.0.18/30
      linkindex: 5
      name: r -> h-3
      neighbors:
      - ifname: eth1
        ipv4: 10.1.0.17/30
        node: h-3
      type: p2p
    - ifindex: 4
      ifname: eth4
      ipv4: 10.1.0.22/30
      linkindex: 6
      name: r -> h-4
      neighbors:
      - ifname: eth1
        ipv4: 10.1.0.21/30
        node: h-4
      type: p2p
    - ifindex: 5
      ifname: eth5
      lag:
        _parentindex: 3
      linkindex: 10
      name: r -> h
      neighbors:
      - ifname: eth3
        node: h
      type: p2p
    - ifindex: 6
      ifname: eth6
      lag:
        _parentindex: 3
      linkindex: 11
      name: r -> h
      neighbors:
      - ifname: eth4
        node: h
      type: p2p
    - bridge: input_12
      ifindex: 7
      ifname: eth7
      linkindex: 12
      name: '[Access VLAN red] r -> h'
      neighbors:
      - ifname: eth5
        ipv4: 172.16.0.10/24
        node: h
      type: lan
      vlan:
        access: red
        access_id: 1000
    - bridge: input_13
      ifindex: 8
      ifname: eth8
      linkindex: 13
      name: '[Access VLAN red] r -> h-2'
      neighbors:
      - ifname: eth3
        ipv4: 172.16.0.11/24
        node: h-2
      type: lan
      vlan:
        access: red
        access_id: 1000
    - bridge: input_14
      ifindex: 9
      ifname: eth9
      linkindex: 14
      name: '[Access VLAN red] r -> h-3'
      neighbors:
      - ifname: eth3
        ipv4: 172.16.0.12/24
        node: h-3
      type: lan
      vlan:
        access: red
        access_id: 1000
    - bridge: input_15
      ifindex: 10
      ifname: eth10
      linkindex: 15
      name: '[Access VLAN red] r -> h-4'
      neighbors:
      - ifname: eth3
        ipv4: 172.16.0.13/24
        node: h-4
      type: lan
      vlan:
        access: red
        access_id: 1000
    - bridge_group: 1
      ifindex: 11
      ifname: vlan1000
      ipv4: 172.16.0.1/24
      name: VLAN red (1000) -> [h,h-2,h-3,h-4]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.10/24
        node: h
      - ifname: vlan1000
        ipv4: 172.16.0.11/24
        node: h-2
      - ifname: vlan1000
        ipv4: 172.16.0.12/24
        node: h-3
      - ifname: vlan1000
        ipv4: 172.16.0.13/24
        node: h-4
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    - ifindex: 12
      ifname: eth12
      linkindex: 0
      neighbors: []
      remote_id: 1
      remote_ifindex: 666
      type: p2p
    - ifindex: 13
      ifname: eth13
      linkindex: 0
      neighbors: []
      remote_id: 1
      remote_ifindex: 666
      type: p2p
    - ifindex: 14
      ifname: eth14
      linkindex: 0
      neighbors: []
      remote_id: 1
      remote_ifindex: 666
      type: p2p
    - ifindex: 15
      ifname: eth15
      linkindex: 0
      neighbors: []
      remote_id: 1
      remote_ifindex: 666
      type: p2p
    - ifindex: 16
      ifname: eth16
      linkindex: 0
      neighbors: []
      remote_id: 1
      remote_ifindex: 666
      type: p2p
    - ifindex: 17
      ifname: eth17
      linkindex: 0
      neighbors: []
      remote_id: 1
      remote_ifindex: 666
      type: p2p
    - ifindex: 18
      ifname: eth18
      linkindex: 0
      neighbors: []
      remote_id: 1
      remote_ifindex: 666
      type: p2p
    - ifindex: 19
      ifname: eth19
      linkindex: 0
      neighbors: []
      remote_id: 1
      remote_ifindex: 666
      type: p2p
    - ifindex: 20
      ifname: eth20
      ipv4: 10.1.0.6/30
      linkindex: 2
      name: r -> h
      neighbors:
      - ifname: eth2
        ipv4: 10.1.0.5/30
        node: h
      type: p2p
    - ifindex: 21
      ifname: eth21
      ipv4: 10.1.0.26/30
      linkindex: 7
      name: r -> h-2
      neighbors:
      - ifname: eth2
        ipv4: 10.1.0.25/30
        node: h-2
      type: p2p
    - ifindex: 22
      ifname: eth22
      ipv4: 10.1.0.30/30
      linkindex: 8
      name: r -> h-3
      neighbors:
      - ifname: eth2
        ipv4: 10.1.0.29/30
        node: h-3
      type: p2p
    - ifindex: 23
      ifname: eth23
      ipv4: 10.1.0.34/30
      linkindex: 9
      name: r -> h-4
      neighbors:
      - ifname: eth2
        ipv4: 10.1.0.33/30
        node: h-4
      type: p2p
    - ifindex: 30000
      ifname: bond1
      ipv4: 10.1.0.9/30
      lag:
        ifindex: 1
        lacp: fast
        lacp_mode: active
        mode: 802.3ad
      linkindex: 3
      name: r -> h
      neighbors:
      - ifname: bond1
        ipv4: 10.1.0.10/30
        node: h
      type: lag
      virtual_interface: true
    lag:
      lacp: fast
      lacp_mode: active
      mode: 802.3ad
    libvirt:
      nic_adapter_count: 24
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.1/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.101
      mac: 08:4f:a9:00:00:01
    module:
    - lag
    - vlan
    name: r
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
plugin:
- repeater
provider: libvirt
vlans:
  red:
    host_count: 0
    id: 1000
    neighbors:
    - ifname: vlan1000
      ipv4: 172.16.0.10/24
      node: h
    - ifname: vlan1000
      ipv4: 172.16.0.1/24
      node: r
    - ifname: vlan1000
      ipv4: 172.16.0.11/24
      node: h-2
    - ifname: vlan1000
      ipv4: 172.16.0.12/24
      node: h-3
    - ifname: vlan1000
      ipv4: 172.16.0.13/24
      node: h-4
    prefix:
      allocation: id_based
      ipv4: 172.16.0.0/24
